File: src/utils/typographyPresets.js
// Per-user typography presets for the rich-text editor.
//
// Data model: the user keeps THREE independent profiles so they can
// save several complete typography setups and switch between them
// with a tab click. Only the ACTIVE profile's values are written to
// :root as CSS variables, so swapping profiles re-renders the editor
// and the view mode with no extra plumbing.
//
//   { active: "profile1",
//     profile1: { p, h1, h2, h3, h4, h5 },
//     profile2: { … },
//     profile3: { … } }
//
// Each block carries five configurable properties:
//   size       — rem string (e.g. "1.5rem")
//   weight     — CSS font-weight number (400 / 500 / 600 / 700)
//   color      — hex / rgb / null (null = inherit)
//   italic     — boolean
//   underline  — boolean

export const DEFAULT_PROFILE = Object.freeze({
  p:  { size: "1rem",     weight: 400, color: null,      italic: false, underline: false },
  h1: { size: "1.75rem",  weight: 800, color: "#4f46e5", italic: false, underline: false },
  h2: { size: "1.5rem",   weight: 700, color: "#059669", italic: false, underline: false },
  h3: { size: "1.25rem",  weight: 600, color: "#0284c7", italic: false, underline: true  },
  h4: { size: "1.125rem", weight: 600, color: "#d97706", italic: true,  underline: false },
  h5: { size: "1rem",     weight: 500, color: "#db2777", italic: true,  underline: true  },
});

export const PROFILE_KEYS = ["profile1", "profile2", "profile3"];
export const DEFAULT_ACTIVE_PROFILE = "profile1";

export const DEFAULT_TYPOGRAPHY_PRESETS = Object.freeze({
  active: DEFAULT_ACTIVE_PROFILE,
  profile1: DEFAULT_PROFILE,
  profile2: DEFAULT_PROFILE,
  profile3: DEFAULT_PROFILE,
});

export const TYPOGRAPHY_SIZE_PRESETS = [
  { value: "0.875rem", label: "14" },
  { value: "1rem",     label: "16" },
  { value: "1.125rem", label: "18" },
  { value: "1.25rem",  label: "20" },
  { value: "1.5rem",   label: "24" },
  { value: "1.75rem",  label: "28" },
  { value: "2rem",     label: "32" },
  { value: "2.25rem",  label: "36" },
  { value: "2.5rem",   label: "40" },
];

export const TYPOGRAPHY_WEIGHT_PRESETS = [
  { value: 400, labelKey: "typographyWeightNormal" },
  { value: 500, labelKey: "typographyWeightMedium" },
  { value: 600, labelKey: "typographyWeightSemibold" },
  { value: 700, labelKey: "typographyWeightBold" },
];

// Swatches offered in the per-block colour picker.
export const TYPOGRAPHY_COLOR_PRESETS = [
  "#111827", "#ef4444", "#f97316", "#eab308",
  "#10b981", "#0ea5e9", "#6366f1", "#a855f7",
  "#ec4899",
];

const BLOCKS = ["p", "h1", "h2", "h3", "h4", "h5"];

const COLOR_RE = /^(#[0-9a-fA-F]{3,8}|rgb|hsl|inherit|transparent)/;

/**
 * Snap a size string to the nearest entry in TYPOGRAPHY_SIZE_PRESETS.
 * The dropdown only offers preset values, so any stored value that
 * isn't in the list (older defaults like "1.35rem" / "1.15rem" / …)
 * would surface as the dropdown showing its first option ("14") while
 * the inline preview rendered the saved size — visually contradicting
 * itself. Snapping on read normalises old data without losing intent.
 */
function nearestPresetSize(value) {
  if (typeof value !== "string") return null;
  const m = value.match(/^([\d.]+)\s*rem$/i);
  if (!m) return null;
  const target = parseFloat(m[1]);
  if (!Number.isFinite(target)) return null;
  let best = TYPOGRAPHY_SIZE_PRESETS[0];
  let bestDiff = Math.abs(parseFloat(best.value) - target);
  for (const p of TYPOGRAPHY_SIZE_PRESETS) {
    const d = Math.abs(parseFloat(p.value) - target);
    if (d < bestDiff) { best = p; bestDiff = d; }
  }
  return best.value;
}

function sanitizeBlock(value, fallback) {
  if (!value || typeof value !== "object") return { ...fallback };
  let size = typeof value.size === "string" ? value.size : fallback.size;
  // Migrate non-preset sizes to the nearest preset entry.
  if (!TYPOGRAPHY_SIZE_PRESETS.some((s) => s.value === size)) {
    const snapped = nearestPresetSize(size);
    if (snapped) size = snapped;
  }
  const weightNum = Number(value.weight);
  const weight = Number.isFinite(weightNum) && weightNum >= 100 && weightNum <= 900
    ? weightNum
    : fallback.weight;
  let color;
  if (value.color === null) color = null;
  else if (typeof value.color === "string" && COLOR_RE.test(value.color)) color = value.color;
  else color = fallback.color;
  const italic = typeof value.italic === "boolean" ? value.italic : fallback.italic;
  const underline = typeof value.underline === "boolean" ? value.underline : fallback.underline;
  return { size, weight, color, italic, underline };
}

function sanitizeProfile(raw, fallback) {
  const input = raw && typeof raw === "object" ? raw : {};
  const out = {};
  for (const k of BLOCKS) {
    out[k] = sanitizeBlock(input[k], fallback[k] || DEFAULT_PROFILE[k]);
  }
  return out;
}

export function normalizeTypographyPresets(raw) {
  const input = raw && typeof raw === "object" ? raw : {};

  // Legacy migration: old shape had the blocks at the top level
  // (no `active` key, no `profile1` key). Move them into profile1.
  const hasNewShape =
    !!input.active ||
    PROFILE_KEYS.some((k) => input[k] && typeof input[k] === "object");
  if (!hasNewShape && (input.p || input.h1 || input.h2 || input.h3 || input.h4 || input.h5)) {
    return {
      active: DEFAULT_ACTIVE_PROFILE,
      profile1: sanitizeProfile(input, DEFAULT_PROFILE),
      profile2: { ...DEFAULT_PROFILE },
      profile3: { ...DEFAULT_PROFILE },
    };
  }

  const active = PROFILE_KEYS.includes(input.active) ? input.active : DEFAULT_ACTIVE_PROFILE;
  return {
    active,
    profile1: sanitizeProfile(input.profile1, DEFAULT_PROFILE),
    profile2: sanitizeProfile(input.profile2, DEFAULT_PROFILE),
    profile3: sanitizeProfile(input.profile3, DEFAULT_PROFILE),
  };
}

/** Return the currently active profile's block map. */
export function getActiveProfile(presets) {
  const n = normalizeTypographyPresets(presets);
  return n[n.active];
}

export function isDefaultTypography(presets) {
  const n = normalizeTypographyPresets(presets);
  if (n.active !== DEFAULT_ACTIVE_PROFILE) return false;
  return PROFILE_KEYS.every((pk) =>
    BLOCKS.every((k) => {
      const a = n[pk][k];
      const b = DEFAULT_PROFILE[k];
      return (
        a.size === b.size &&
        a.weight === b.weight &&
        a.color === b.color &&
        a.italic === b.italic &&
        a.underline === b.underline
      );
    }),
  );
}

/**
 * Push the ACTIVE profile's block styles onto the document root as
 * CSS variables. Switching profiles triggers a fresh call so the UI
 * re-renders with the new look immediately.
 */
export function applyTypographyPresets(presets) {
  if (typeof document === "undefined") return;
  const n = normalizeTypographyPresets(presets);
  const profile = n[n.active];
  const root = document.documentElement;
  for (const k of BLOCKS) {
    const p = profile[k];
    root.style.setProperty(`--gk-type-${k}-size`,      p.size);
    root.style.setProperty(`--gk-type-${k}-weight`,    String(p.weight));
    root.style.setProperty(`--gk-type-${k}-color`,     p.color || "inherit");
    root.style.setProperty(`--gk-type-${k}-italic`,    p.italic ? "italic" : "normal");
    root.style.setProperty(`--gk-type-${k}-underline`, p.underline ? "underline" : "none");
  }
}

export const TYPOGRAPHY_STORAGE_KEY = "typographyPresets";

File: android/app/src/main/java/com/glasskeep/app/nativeapp/data/TypographyPresets.kt
package com.glasskeep.app.nativeapp.data

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Per-user typography presets for rich-text notes, ported from
 * src/utils/typographyPresets.js.
 *
 * The user keeps THREE independent profiles and one `active` pointer, so
 * several complete setups can be saved and swapped with a tap. Each of the
 * six blocks (paragraph, headings 1-5) carries five properties: size,
 * weight, colour, italic and underline. The whole blob is one field of the
 * server-side user settings, so it follows the account across devices,
 * same as the workspace theme.
 *
 * The blob travels as raw JSON: [TypographyPresets.normalize] reads it the
 * way normalizeTypographyPresets() does, which has to tell a `"color": null`
 * (inherit) from a block with no colour at all (its default colour), and
 * [TypographyPresets.toJson] writes every key back, nulls included.
 */
/** One block's five resolved properties. [size] is in rem, the unit the
 *  web stores; multiply by 16 for the sp value (1rem = 16px there). */
data class TypographyBlock(
    val size: Float,
    val weight: Int,
    /** A CSS colour, or null for "inherit" (the note's own text colour). */
    val color: String?,
    val italic: Boolean,
    val underline: Boolean,
)

data class TypographyProfile(
    val p: TypographyBlock,
    val h1: TypographyBlock,
    val h2: TypographyBlock,
    val h3: TypographyBlock,
    val h4: TypographyBlock,
    val h5: TypographyBlock,
) {
    fun forKind(kind: RichBlockKind): TypographyBlock = when (kind) {
        RichBlockKind.HEADING_1 -> h1
        RichBlockKind.HEADING_2 -> h2
        RichBlockKind.HEADING_3 -> h3
        RichBlockKind.HEADING_4 -> h4
        RichBlockKind.HEADING_5 -> h5
        else -> p
    }

    fun withBlock(key: String, block: TypographyBlock): TypographyProfile = when (key) {
        "h1" -> copy(h1 = block)
        "h2" -> copy(h2 = block)
        "h3" -> copy(h3 = block)
        "h4" -> copy(h4 = block)
        "h5" -> copy(h5 = block)
        else -> copy(p = block)
    }

    fun byKey(key: String): TypographyBlock = when (key) {
        "h1" -> h1
        "h2" -> h2
        "h3" -> h3
        "h4" -> h4
        "h5" -> h5
        else -> p
    }
}

data class TypographyPresets(
    val active: String,
    val profile1: TypographyProfile,
    val profile2: TypographyProfile,
    val profile3: TypographyProfile,
) {
    val activeProfile: TypographyProfile
        get() = profileFor(active)

    fun profileFor(key: String): TypographyProfile = when (key) {
        "profile2" -> profile2
        "profile3" -> profile3
        else -> profile1
    }

    fun withProfile(key: String, profile: TypographyProfile): TypographyPresets = when (key) {
        "profile2" -> copy(profile2 = profile)
        "profile3" -> copy(profile3 = profile)
        else -> copy(profile1 = profile)
    }

    fun toJson(): JsonObject = buildJsonObject {
        put("active", active)
        for (key in PROFILE_KEYS) put(key, profileFor(key).toJson())
    }

    val isDefault: Boolean
        get() = active == DEFAULT_ACTIVE &&
            profile1 == DEFAULT_PROFILE && profile2 == DEFAULT_PROFILE && profile3 == DEFAULT_PROFILE

    companion object {
        const val DEFAULT_ACTIVE = "profile1"

        val PROFILE_KEYS = listOf("profile1", "profile2", "profile3")
        val BLOCK_KEYS = listOf("p", "h1", "h2", "h3", "h4", "h5")

        /** DEFAULT_PROFILE (typographyPresets.js:21-28), value for value. */
        val DEFAULT_PROFILE = TypographyProfile(
            p = TypographyBlock(1f, 400, null, italic = false, underline = false),
            h1 = TypographyBlock(1.75f, 800, "#4f46e5", italic = false, underline = false),
            h2 = TypographyBlock(1.5f, 700, "#059669", italic = false, underline = false),
            h3 = TypographyBlock(1.25f, 600, "#0284c7", italic = false, underline = true),
            h4 = TypographyBlock(1.125f, 600, "#d97706", italic = true, underline = false),
            h5 = TypographyBlock(1f, 500, "#db2777", italic = true, underline = true),
        )

        val DEFAULT = TypographyPresets(DEFAULT_ACTIVE, DEFAULT_PROFILE, DEFAULT_PROFILE, DEFAULT_PROFILE)

        /** TYPOGRAPHY_SIZE_PRESETS, in rem. */
        val SIZE_PRESETS = listOf(0.875f, 1f, 1.125f, 1.25f, 1.5f, 1.75f, 2f, 2.25f, 2.5f)

        /** TYPOGRAPHY_WEIGHT_PRESETS. */
        val WEIGHT_PRESETS = listOf(400, 500, 600, 700)

        /** TYPOGRAPHY_COLOR_PRESETS: the nine swatches the block colour
         *  picker offers, on top of "inherit". */
        val COLOR_PRESETS = listOf(
            "#111827", "#ef4444", "#f97316", "#eab308",
            "#10b981", "#0ea5e9", "#6366f1", "#a855f7",
            "#ec4899",
        )

        /** normalizeTypographyPresets(): the legacy shape, the six blocks
         *  at the top level with no profiles at all, becomes profile 1. */
        fun normalize(raw: JsonElement?): TypographyPresets {
            val input = raw as? JsonObject ?: return DEFAULT
            val hasNewShape = input["active"].isTruthy() || PROFILE_KEYS.any { input[it] is JsonObject }
            if (!hasNewShape && BLOCK_KEYS.any { input[it].isTruthy() }) {
                return TypographyPresets(DEFAULT_ACTIVE, sanitize(input), DEFAULT_PROFILE, DEFAULT_PROFILE)
            }
            return TypographyPresets(
                active = input.string("active")?.takeIf { it in PROFILE_KEYS } ?: DEFAULT_ACTIVE,
                profile1 = sanitize(input["profile1"]),
                profile2 = sanitize(input["profile2"]),
                profile3 = sanitize(input["profile3"]),
            )
        }

        private fun sanitize(raw: JsonElement?): TypographyProfile {
            val input = raw as? JsonObject
            return TypographyProfile(
                p = sanitizeBlock(input?.get("p"), DEFAULT_PROFILE.p),
                h1 = sanitizeBlock(input?.get("h1"), DEFAULT_PROFILE.h1),
                h2 = sanitizeBlock(input?.get("h2"), DEFAULT_PROFILE.h2),
                h3 = sanitizeBlock(input?.get("h3"), DEFAULT_PROFILE.h3),
                h4 = sanitizeBlock(input?.get("h4"), DEFAULT_PROFILE.h4),
                h5 = sanitizeBlock(input?.get("h5"), DEFAULT_PROFILE.h5),
            )
        }

        private fun sanitizeBlock(raw: JsonElement?, fallback: TypographyBlock): TypographyBlock {
            val value = raw as? JsonObject ?: return fallback
            return TypographyBlock(
                size = value.string("size")?.let(::presetSize) ?: fallback.size,
                weight = (value["weight"] as? JsonPrimitive)?.content?.trim()?.toDoubleOrNull()
                    ?.takeIf { it in 100.0..900.0 }?.toInt() ?: fallback.weight,
                color = if (value["color"] is JsonNull) {
                    null
                } else {
                    value.string("color")?.takeIf { ColorRegex.containsMatchIn(it) } ?: fallback.color
                },
                italic = value.boolean("italic") ?: fallback.italic,
                underline = value.boolean("underline") ?: fallback.underline,
            )
        }

        private val ColorRegex = Regex("^(#[0-9a-fA-F]{3,8}|rgb|hsl|inherit|transparent)")
        private val RemRegex = Regex("""^([\d.]+)\s*rem$""", RegexOption.IGNORE_CASE)

        /** Snaps a stored "1.35rem" to the nearest offered preset, exactly
         *  as nearestPresetSize() does on the web: the picker only offers
         *  the preset list, so an off-list value would otherwise show one
         *  size in the dropdown and render another. */
        private fun presetSize(raw: String): Float? {
            val value = RemRegex.find(raw)?.groupValues?.get(1)?.toFloatOrNull() ?: return null
            return SIZE_PRESETS.minByOrNull { kotlin.math.abs(it - value) }
        }
    }
}

private fun TypographyProfile.toJson() = buildJsonObject {
    for (key in TypographyPresets.BLOCK_KEYS) put(key, byKey(key).toJson())
}

private fun TypographyBlock.toJson() = buildJsonObject {
    put("size", "${formatRem(size)}rem")
    put("weight", weight)
    put("color", color)
    put("italic", italic)
    put("underline", underline)
}

private fun JsonObject.string(key: String): String? =
    (this[key] as? JsonPrimitive)?.takeIf { it.isString }?.content

private fun JsonObject.boolean(key: String): Boolean? =
    (this[key] as? JsonPrimitive)?.takeIf { !it.isString }?.booleanOrNull

/** JavaScript truthiness, for the legacy-shape test. */
private fun JsonElement?.isTruthy(): Boolean = when (this) {
    null, JsonNull -> false
    is JsonPrimitive -> if (isString) content.isNotEmpty() else content != "false" && content.toDoubleOrNull() != 0.0
    else -> true
}

/** "1rem", "1.125rem": the web writes plain decimals, never "1.0rem". */
private fun formatRem(value: Float): String {
    val rounded = Math.round(value * 1000f) / 1000f
    return if (rounded == rounded.toInt().toFloat()) rounded.toInt().toString() else rounded.toString()
}
